runtime/parachains/src/paras_inherent.rs

```diff
@@ -105,32 +181,154 @@ pub mod pallet {
 		const INHERENT_IDENTIFIER: InherentIdentifier = PARACHAINS_INHERENT_IDENTIFIER;
 
 		fn create_inherent(data: &InherentData) -> Option<Self::Call> {
-			let mut inherent_data: ParachainsInherentData<T::Header> =
-				match data.get_data(&Self::INHERENT_IDENTIFIER) {
-					Ok(Some(d)) => d,
-					Ok(None) => return None,
-					Err(_) => {
-						log::warn!(target: LOG_TARGET, "ParachainsInherentData failed to decode");
-
-						return None
-					},
-				};
+			let parent_hash = <frame_system::Pallet<T>>::parent_hash();
+
+			let ParachainsInherentData::<T::Header> {
+				bitfields,
+				backed_candidates,
+				mut disputes,
+				parent_header,
+			} = match data.get_data(&Self::INHERENT_IDENTIFIER) {
+				Ok(Some(d)) => d,
+				Ok(None) => return None,
+				Err(_) => {
+					log::warn!(target: LOG_TARGET, "ParachainsInherentData failed to decode");
+					return None
+				},
+			};
+
+			if parent_hash != parent_header.hash() {
+				log::warn!(
+					target: LOG_TARGET,
+					"ParachainsInherentData references a different parent header hash than frame"
+				);
+				return None
+			}
+
+			let current_session = <shared::Pallet<T>>::session_index();
+			let expected_bits = <scheduler::Pallet<T>>::availability_cores().len();
+			let validator_public = shared::Pallet::<T>::active_validator_keys();
+
+			T::DisputesHandler::filter_multi_dispute_data(&mut disputes);
+
+			let scheduled: Vec<CoreAssignment> = <scheduler::Pallet<T>>::scheduled();
+
+			let (freed_cores, concluded_invalid_disputes) =
+				frame_support::storage::with_transaction(|| {
+					// we don't care about fresh or not disputes
+					// this writes them to storage, so let's query it via those means
+					// if this fails for whatever reason, that's ok
```

Comment:
I assume if anything returns an error this closure returns `TransactionOutcome::Rollback` ? Its just not immediatley obvious to me how we go from a normal error to `TransactionOutcome::Rollback` but I assume there is a From blanket impl or something for errors?

Reply:
There are no `?` in that closure. All errors are just printed.